pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -138,6 +138,37 @@ func (t ResolvedPipelineTask) IsCustomTask() bool {
 	return t.CustomTask
 }
 
+// getReason returns the latest reason if the run has completed successfully
+// If the PipelineTask has a Matrix, getReason returns the failure reason for any failure
+// otherwise, it returns Succeeded
+func (t ResolvedPipelineTask) getReason() string {
+	if t.IsCustomTask() {
+		if len(t.CustomRuns) == 0 {
+			return ""
+		}
+		for _, run := range t.CustomRuns {
+			if !run.IsSuccessful() && len(run.Status.Conditions) >= 1 {
+				return run.Status.Conditions[0].Reason
+			}
+		}
+		if len(t.CustomRuns) >= 1 && len(t.CustomRuns[0].Status.Conditions) >= 1 {
+			return t.CustomRuns[0].Status.Conditions[0].Reason
+		}
+	}
+	if len(t.TaskRuns) == 0 {
+		return ""
+	}
+	for _, taskRun := range t.TaskRuns {
+		if !taskRun.IsSuccessful() && len(taskRun.Status.Conditions) >= 1 {
+			return taskRun.Status.Conditions[0].Reason
+		}
+	}
+	if len(t.TaskRuns) >= 1 && len(t.TaskRuns[0].Status.Conditions) >= 1 {
+		return t.TaskRuns[0].Status.Conditions[0].Reason
+	}
+	return ""
```

Comment:
NIT: This line is not covered by testing, I think we only get here if `len(t.TaskRuns) >= 1` and `len(t.TaskRuns[0].Status.Conditions) == 0`, which I think will only happen if the `TaskRun` was created but never executed. Would it make sense to add this case to the unit test?

Reply:
It makes sense to add a unit test. However, this code path isn't typically reachable in practice because `.getReason` and similar functions are only called from the finally tasks.

Reply:
Unit test: https://github.com/tektoncd/pipeline/pull/8127/files#diff-087040c828aabacd714f2761d662149b1f728af59ce14d04e83f39cba2f059f0R4824-R4832